modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/processing/BlockAcceptanceLogic.scala

```diff
@@ -0,0 +1,132 @@
+                      .modify(_.updated(parent, parentUsageCount |+| usageIncrement))
+                    val updatedUsageCount =
+                      if (parentUsageCount >= deprecationThreshold)
+                        usageCount |+| usageIncrement
+                      else
+                        usageCount
+                    (updatedState, updatedUsageCount)
+                  }
+                  .toRight[NotAcceptedBlock](InvalidBlock(blockRef, ParentNotFound(parent)))
+                  .toEitherT[F]
+            }
+          }
+      } yield (updatedState, usageCount)
+
+    private def processTransactions(
+      signedBlock: Signed[DAGBlock],
+      state: BlockAcceptanceState
+    ): BlockAcceptingM[F, Unit] =
+      for {
+        blockRef <- EitherT.liftF(BlockReference.of(signedBlock))
+
+        (updatedState, _) <- signedBlock.transactions.toList.sorted
+          .foldLeft(rightT(state, none[BlockAwaitReason])) { (acc, tx) =>
+            {
+              EitherT.liftF(TransactionReference.of(tx)).flatMap { txRef =>
+                acc.flatMap {
+                  case (state, maybeNotReady) =>
+                    val lastTxRef = state.lastTxRefs.getOrElse(tx.source, TransactionReference.empty)
+
+                    if (tx.parent.ordinal < lastTxRef.ordinal) {
+                      leftT(InvalidBlock(blockRef, InvalidTransaction(txRef, OrdinalTooLow(tx.ordinal))))
+                    } else if (tx.parent.ordinal > lastTxRef.ordinal) {
+                      // we don't want to "fail fast" yet, therefore `rightT` is used, even though we could use `leftT`
+                      // this way we make sure that the block is validated before we classify it as awaiting
+                      rightT(state, maybeNotReady.orElse(AwaitingTransaction(txRef, OrdinalTooHigh(tx.ordinal)).some))
+                    } else {
+                      if (tx.parent.hash =!= lastTxRef.hash) {
+                        leftT(InvalidBlock(blockRef, InvalidTransaction(txRef, IncorrectParent(tx.parent))))
+                      } else {
+                        val balanceErrorOrState = for {
+                          initSrcBalance <- state.balances.getOrElse(tx.source, Balance.empty).minus(tx.amount)
+                          srcBalance <- initSrcBalance.minus(tx.fee)
+                          dstBalance <- state.balances.getOrElse(tx.destination, Balance.empty).plus(tx.amount)
+
+                          balances = state.balances.updated(tx.source, srcBalance).updated(tx.destination, dstBalance)
+                          lastTxRefs = state.lastTxRefs.updated(tx.source, txRef)
+                        } yield state.copy(balances = balances, lastTxRefs = lastTxRefs)
+
+                        balanceErrorOrState
+                          .map(state => (state, maybeNotReady))
+                          .leftMap(err => invalidBlock(blockRef, InvalidTransaction(txRef, InvalidBalance(err))))
+                          .toEitherT[F]
```

Comment:
I'd say that this kind of validation will be hard to maintain. I'd use `Validation` monad, extract separate pieces into own methods (like checking the parent hash / reference, checking the ordinal, etc) and I'd highly opt to reuse the transaction validator. Then, you can use tx validator to obtain whether tx is valid/invalid and add a custom validation for tips if needed here.